---
title: "bmap_description"
author: "Roan"
date: "7/26/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/home/rstudio/")
pdf.options(paper = "a4")

```

```{r}
if (!require("pacman")) install.packages("pacman")
pacman::p_load(tidyverse,ggrepel,vroom,ggpubr,dplyr,ggplot2,reshape2,scales,plyr,
Hmisc)
```

#VIOLIN PLOTS
```{r}
bmap<-vroom("./data/blood_grn_sig.txt.gz", delim="\t")
#just to make sure no duplicates
bmap<-distinct(bmap) %>% 
  dplyr::select(gencode_id, snp, interaction_type, snp_chr, gene) %>% mutate(ensembl_gene_id=str_extract(bmap$gencode_id, regex("(ENSG[0-9]{11})")))

bmap<- bmap %>% mutate("eQTL_eGene" = paste0(bmap$snp,"_", bmap$ensembl_gene_id))

```

#find n_eQTLs per gene per interaction type
```{r}
bmap_inter<- bmap%>% 
  filter(interaction_type=="Trans-interchromosomal") %>% 
   dplyr::count(ensembl_gene_id, name="n_eQTLs") %>% 
  mutate("Key"="Trans-inter")

bmap_intra <-bmap%>% 
  filter(interaction_type=="Trans-intrachromosomal") %>% 
     dplyr::count(ensembl_gene_id, name="n_eQTLs") %>% 
  mutate("Key"="Trans-intra")

bmap_cis<-bmap %>% 
  filter(interaction_type=="Cis") %>% 
     dplyr::count(ensembl_gene_id, name="n_eQTLs") %>% 
  mutate("Key"="Cis")

bmap_combined <- rbind(bmap_inter, bmap_intra) %>% 
  rbind(bmap_cis)

 bmap_combined<-bmap_combined %>% 
   mutate(Key=factor(Key, levels=c("Cis", "Trans-intra","Trans-inter"))) 

```


#stat test (manual)
```{r}
bmap_combined %>% group_by(Key) %>% dplyr::summarise(mean=mean(log10(n_eQTLs)))
```
```{r}
bmap_combined %>% group_by(Key) %>% dplyr::summarise(sd=sd(log10(n_eQTLs)))
```

#stat test (plot)
```{r}
my_comparisons <- list( c("Cis", "Trans-intra"), c("Cis", "Trans-inter"), c("Trans-inter", "Trans-intra") )
```

#plot
```{r}
Figure2a<-ggplot(bmap_combined, aes(x=Key, y=log10(n_eQTLs), fill=Key)) + 
    geom_violin(trim=T)  + theme_minimal() + stat_summary(fun.y = "median", geom = "point", shape = 3, size = 4, color = "black") +  stat_summary(fun.data=mean_sdl, fun.args = list(mult = 1), 
                 geom="pointrange", color="brown", shape=3, size=0.85) + 
  labs(y = "log10(Number of eQTLs)") +
  scale_fill_brewer(palette="RdBu") +  theme(legend.position = "none") + theme(axis.text.x= element_text(size=12)) +
   stat_compare_means(comparisons = my_comparisons, method = "t.test", label = "p.signif") 
```

```{r}
#save as pdf
# pdf("./plots/Figure2a.pdf", height=4, width=5, paper = "USr")
# print(Figure2a)
# dev.off()
```

#median tpm measured from blood (GTEx)
```{r}
tpm<-vroom("./data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct", delim = "\t", skip=2, col_select = c("Description", "Whole Blood", "Name")) %>% 
  dplyr::rename("gencode_id"= Name)
tpm<- tpm %>% mutate(ensembl_gene_id=str_extract(tpm$gencode_id, regex("(ENSG[0-9]{11})")))
```

#annotate tpm per gene
```{r}
bmap_combined_tpm<-merge(bmap_combined, tpm, by= "ensembl_gene_id")
suppl_table_2<-bmap_combined_tpm %>% dplyr::rename("gene"="Description", "interaction"="Key", "whole_blood_median_tpm"="Whole Blood") %>% dplyr::relocate(gene, ensembl_gene_id, gencode_id,whole_blood_median_tpm, interaction) %>% dplyr::select(-n_eQTLs, -interaction) %>% distinct()

# write.table(suppl_table_2, "./results/bmap_genes_expression_levels.txt", quote=F, row.names=F, sep="\t")
```

#stat test 
```{r}
bmap_combined_tpm%>% group_by(Key) %>% dplyr::summarise(mean=mean((`Whole Blood`)))
```
```{r}
bmap_combined_tpm %>% group_by(Key) %>% dplyr::summarise(mean=mean(`Whole Blood`))
```

#add +1 to allow log transformation of zeros and plot
```{r}
Figure2b<-ggplot(bmap_combined_tpm, aes(x=Key, y=log10(`Whole Blood` +1), fill=Key)) + 
    geom_violin(trim=T)  + theme_minimal() + stat_summary(fun.y = "median", geom = "point", shape = 3, size = 4, color = "black") + 
 stat_summary(fun.data=mean_sdl, fun.args = list(mult = 1),  
                 geom="pointrange", color="brown", shape=3, size=0.85)+   labs(y = "log10(median TPM)") +
  scale_fill_brewer(palette="RdBu")  + theme(legend.position = "none") + theme(axis.text.x= element_text(size=12)) +
   stat_compare_means(comparisons = my_comparisons, method = "t.test", label = "p.signif") 

```
#save as pdf
```{r}
# pdf("./plots/Figure2b.pdf", height=4, width=5, paper = "USr")
# print(Figure2b)
# dev.off()
```

#Barplot gtex codes 3D
#PROPRTION GTEX CODES3D 
#read annotated GTEx file including both cis and trans data
```{r}
#read cis eQTLs
gtex_sig_cis<- read.table("./data/annotated_GTEx_v8_sig_WBlood_CIS_SNPs.txt.gz", sep ="\t", row.names = 1) %>% dplyr::select(gene_id, rs_id_dbSNP151_GRCh38p7, variant_id) %>% mutate(interaction_type="cis")

#read trans eQTLs (all are trans intra) 
gtex_sig_trans<- read.table("./data/annotated_GTEx_v8_sig_WBlood_TRANS_SNPs.txt", sep="\t", row.names= 1) %>% dplyr::select(gene_id, rs_id_dbSNP151_GRCh38p7, variant_id) %>% mutate(interaction_type="trans_intrachromosomal")

gtex_sig<- rbind(gtex_sig_trans, gtex_sig_cis)
rm(gtex_sig_cis)

gtex_sig <- gtex_sig %>% mutate(ensembl_gene_id=str_extract(gtex_sig$gene_id, regex("(ENSG[0-9]{11})")))

gtex_sig <- gtex_sig %>% mutate("eQTL_eGene" = paste0(gtex_sig$rs_id_dbSNP151_GRCh38p7,"_", gtex_sig$ensembl_gene_id))
```

#table listing interaction types and their numbers 
```{r}
Key<-c("Total GTEx", 
       "Total Blood GRN",
       "Blood GRN & GTEx", 
       "Blood GRN Trans-intra",
       "Blood GRN Trans-inter",
       "GTEx Trans-intra")

eQTL_eGene<-c(nrow(gtex_sig),
             nrow(bmap),
             nrow(merge(gtex_sig, bmap, by="eQTL_eGene")),
             nrow(bmap %>% filter(interaction_type == "Trans-intrachromosomal" )),
             nrow(bmap %>% filter(interaction_type ==  "Trans-interchromosomal")),
             nrow(gtex_sig_trans)) #GTEx does not have trans-inter interactions 

table_gtex_bmap<-data.frame(Key, eQTL_eGene)
#"vjust" to specify the label position, alpha inside geom_bar sets transparency
```

#plot significant eQTL-eGene interactions identified by CoDeS3D and GTEx (v8) in whole blood 
```{r}
FigureS1a<-ggplot(data=table_gtex_bmap, aes(x= Key, y=eQTL_eGene, fill=Key)) + 
  geom_bar(stat="identity", width=0.6, alpha=0.8, position = position_dodge(1)) +
  geom_text(aes(label=eQTL_eGene), vjust=-0.2,
            color="black", size=3) +
  theme_classic() + theme(legend.position = "none")+
  theme(plot.title = element_text(face="bold",size=9, hjust = 0.5)) +
  theme(axis.text.x=element_text(angle=40,hjust=1,vjust=1)) +
  scale_x_discrete(limits=c("Total GTEx", "Total Blood GRN", "Blood GRN & GTEx", "Blood GRN Trans-intra", "Blood GRN Trans-inter","GTEx Trans-intra")) +
  scale_fill_manual(values=c('antiquewhite','darkorange4', 'black','goldenrod3', "black", "orange2")) + scale_y_continuous(labels = function(x) format(x, scientific = TRUE))
```

#save as pdf
```{r}
# pdf("./plots/FigureS1a.pdf", height=5, width=6, paper = "USr")
# print(FigureS1a)
# dev.off()
```

#prepare table for proprtion bar plot
```{r}
   n_gtex_eqtl_egene_by_chr <- gtex_sig %>% mutate(snp_chr =str_extract(gtex_sig$variant_id, regex("(^chr[A-Z0-9]*)"))) %>% dplyr::select(snp_chr, interaction_type, variant_id, eQTL_eGene) %>% distinct() %>% dplyr::count(snp_chr) %>% mutate(Key= "GTEx")
  
 n_bmap_eqtl_egene_by_chr <- bmap %>% dplyr::select(snp_chr, interaction_type, snp, gencode_id) %>% distinct() %>%  dplyr::count(snp_chr) %>% mutate(Key="Blood GRN")
 
 overlap <- merge(bmap, gtex_sig, by="eQTL_eGene") %>% dplyr::select(snp_chr, variant_id, snp, gencode_id) %>% distinct() %>%  dplyr::count(snp_chr) %>% mutate(Key="Blood GRN & GTEx")
 
 #"eQTL_eGene_interactions" col will be used to compute percentage
 proportion_table <- rbind(n_gtex_eqtl_egene_by_chr, n_bmap_eqtl_egene_by_chr, overlap) %>% dplyr::rename("Chromosome" = snp_chr, "eQTL_eGene_interactions"=n) 
 
 proportion_table <- proportion_table  %>% mutate("Interactions" = eQTL_eGene_interactions)  %>% mutate(Chromosome = gsub("chr", "", proportion_table$Chromosome)) %>% arrange(Chromosome) 
 
 #change levels so that the interactions are ordered smallest to largest (else cumulative sum will be miss calculated)
proportion_table <- proportion_table %>%
      mutate(Key=factor(proportion_table$Key, levels = c("GTEx", "Blood GRN","Blood GRN & GTEx")))
 
```

#plot
```{r}
#expand 0,0 forces Y axis to start at origin
FigureS1b<-ggplot(proportion_table) +
  geom_bar(aes(x = Chromosome, y = eQTL_eGene_interactions, fill=Key), stat="identity", position="fill", color = "black") +
  geom_point(aes(x = Chromosome, y = Interactions/400000, shape=Key),size=2.7, color="firebrick4") +
  scale_y_continuous(name= "% eQTL-eGene interactions",labels = percent_format(), expand = c(0, 0),
                     sec.axis = sec_axis(~.*400000, name = "Number of eQTL-eGene interactions")) +
  scale_fill_manual(values=c( "#f7f7f7", "#E69F00", 'antiquewhite')) +
  theme_minimal()+ theme(plot.title = element_text(face="bold", hjust = 0.5))+
  scale_x_discrete(limits=c("1","2","3","4","5","6","7","8","9","10", "11", "12",
                            "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "X")) +
  theme(
    axis.title.y = element_text(color = "black"),
    axis.text.y = element_text(color = "black"),
    axis.title.y.right = element_text(color = "firebrick4"),
    axis.text.y.right = element_text(color = "firebrick4")
  ) +labs(fill="% eQTL-eGene interactions",shape="Number of eQTL-eGene interactions")
```

#save as pdf
```{r}
# pdf("./plots/FigureS1b.pdf", height=6, width=10, paper = "USr")
# print(FigureS1b)
# dev.off()
```


#LOEUF SCORE DENSITY PLOT 
```{r}
categories<-bmap %>% dplyr::select(interaction_type, gene) %>% 
  distinct() %>% 
  mutate(code=1) %>% 
  pivot_wider(names_from = interaction_type, values_from= code) %>% 
  mutate(interaction = case_when(
    Cis==1 & is.na(`Trans-interchromosomal`) & is.na(`Trans-intrachromosomal`) ~ "cis", 
     is.na(Cis) & `Trans-interchromosomal`==1 & is.na(`Trans-intrachromosomal`) ~ "trans-inter",
    is.na(Cis) & is.na(`Trans-interchromosomal`) & `Trans-intrachromosomal`==1 ~ "trans-intra", 
        Cis==1 & `Trans-interchromosomal`==1  & is.na(`Trans-intrachromosomal`) ~ "cis_inter", 
     Cis==1 & is.na(`Trans-interchromosomal`) & `Trans-intrachromosomal`==1 ~ "cis_intra",
    is.na(Cis) & `Trans-interchromosomal`==1 & `Trans-intrachromosomal`==1 ~ "inter_intra", 
    Cis==1 & `Trans-interchromosomal`==1 & `Trans-intrachromosomal`==1 ~ "cis_inter_intra" 
    
  )
  ) %>% dplyr::select(gene, interaction)

categories<- categories %>%  dplyr::count(interaction) %>% merge(categories, by="interaction")

```


```{r}
#extract gene name and pli/loeuff score from gnomad table to make gnomad constraint table
constraint_table<-vroom("./data/gnomad.v2.1.1.lof_metrics.by_gene.txt", delim = "\t", col_select = c("gene", "oe_lof_upper"))

# merge(constraint_table, categories, by="gene") %>% dplyr::select(-n) %>%  write.table( "./results/bmap_loeuf_scores.txt", quote=F, row.names=F, sep="\t")

categories<- categories %>%  mutate(Interaction_type=paste0(categories$interaction, " ", "(", categories$n, ")")) %>% dplyr::select(Interaction_type, "gene" ) %>% merge(constraint_table, by="gene") %>% mutate(Interaction_type =as.factor(Interaction_type))

```
#plot 
```{r}
#SHOW GUIDE-FALSE GETS RID OF BOX THAT SURROUNDS LINE IN THE LEGEND
Figure2c<-ggplot(categories, aes(oe_lof_upper, color=Interaction_type)) +
  labs(x= "LOEUF", y='Density') +
  geom_density( alpha = 0.1, show_guide=FALSE, size=0.8) + theme_classic() +
  scale_color_manual(breaks=c("inter_intra (541)",
                              "cis_inter_intra (1781)",
                              "trans-intra (1159)",
                              "trans-inter (194)",
                              "cis_inter (474)",
                              "cis_intra (3350)",
                              "cis (7363)"),
  values=c('lightsalmon4', "#E69F00", "firebrick", "black",
           "darkslategray4", '#999999', "slateblue4")) +   scale_y_continuous(limit= c(0,1.5)) +
  guides(color=guide_legend("Spatial genes")) + theme(
    legend.title = element_text(size = 12, face="bold") ,
    legend.text = element_text(size = 10),
    plot.title = element_text(size= 12, hjust = 0))  + 
  stat_density(aes(x=oe_lof_upper), geom="line",position="identity")
```

```{r}
#save as pdf
pdf("./plots/Figure2c.pdf", height=4, width=7, paper = "USr")
print(Figure2c)
dev.off()
```


#correlation plot for all snps in GTEx vs blood map snps 
#READING SIG BLOOD MAP EQTLS
```{r}
blood_map_eqtls_all_count <- bmap %>% dplyr::count(snp_chr)

#All whole blood SNPs from GTEX
gtex_snps<-vroom("./data/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.lookup_table_shortened.txt.gz", delim ="\t", col_select = c(variant_id, chr))

#counts the number of SNPs per chromosome 
gtex_snps_count<-gtex_snps %>% dplyr::count(chr)
```

#calculate gene density per Mb for each chr
```{r}
genomeSum<-read.csv("./data/genomeSummary_grch38p13.txt", sep="\t", skip=1)
genomeSum <- genomeSum %>% 
  mutate(geneDensity= genomeSum$Gene/genomeSum$Size..Mb.) %>% #compute gene density
  mutate(Chromosome= tolower(Type)) %>% dplyr::select(geneDensity, Chromosome) #make chromosome name lower case
```


#MERGE TWO COUNT TABLES AND GENE DENSITY INFORMATION PER CHR
```{r}
blood_map_count_merged<-merge(blood_map_eqtls_all_count, gtex_snps_count, by=1, all=TRUE ) %>% 
  dplyr::rename("Chromosome"=snp_chr, "eQTL_eGene_interactions"=n.x, "num_snps"=n.y) %>%
  merge(genomeSum, by="Chromosome")
```

#plot
```{r}
#The limits argument is set to -100000 so that the CI band can extend to the end of the plot
Figure2d<-ggplot(blood_map_count_merged, aes(x=num_snps, y= eQTL_eGene_interactions)) + geom_smooth(fullrange=TRUE, method = lm, alpha=0.2, level=0.95, color="firebrick") +
  geom_count(aes(size=geneDensity))+
  #repel adds labels to points, segment length specifies when a line should connect a label and a point
  geom_text_repel(aes(label = Chromosome),
                  min.segment.length = 0,
                  box.padding   = 0.35,
                  point.padding = 0.5,
                  segment.color = 'grey50') +
  #stat cor adds R and pval to plot. .x and .y specify position of label in plot
  stat_cor(method = "pearson", label.x = 3, label.y = 220000, size=3.2)+
  theme_minimal() +
  #expand=c(0,0) makes sure axix starts at origin
  scale_x_continuous("Number of snps", expand = c(0, 0), limits = c(0, NA)) +
  #setting limits to -big number ensures the CI band extends to ends of plot along with coord_cartesian
  scale_y_continuous("Number of spatial eQTL_eGene_interactions", expand = c(0, 0), limits = c(-1000000, 230000)) +
  labs(title = "Correlation of spatial blood map eQTLs and snps reported in GTEx", hjust=1) +
  coord_cartesian(xlim=c(0,NA), ylim=c(0,NA)) + guides(size=guide_legend("Genes /Mb")) +
# to set limits of geom_count point size 
 scale_size_continuous(limits = c(10,45))
```

#save as pdf
```{r}
# pdf("./plots/Figure2d.pdf", height=4, width=7, paper = "USr")
# print(Figure2d)
# dev.off()
```


