---
title: "bmap_description"
author: "Roan"
date: "3/9/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/home/rstudio/")
pdf.options(paper = "a4")
```

```{r}
if (!require("pacman")) install.packages("pacman")
pacman::p_load(tidyverse,vroom,gprofiler2,ggpubr,Hmisc)
```

```{r}
tpm_0.1<-vroom("./data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct", delim = "\t", skip=2, col_select = c("Description", "Whole Blood", "Name")) %>% dplyr::rename("gencode_id" = Name, "gene" = Description) %>% filter(`Whole Blood`> 0.1) %>% distinct()

tpm_0.1 <- tpm_0.1 %>% mutate(ensembl_gene_id=str_extract(tpm_0.1$gencode_id, regex("(ENSG[0-9]{11})")))
```

```{r}
bmap<-vroom("./data/blood_grn_sig.txt.gz", delim="\t", col_select = c(gene, gencode_id, expression)) %>% distinct() 
```

#find bmap genes that are expressed in blood >0.1 (these are the ones shared with tpm 0.1)
```{r}
expressed_bmap_genes <- bmap %>% filter(expression > 0.1) %>% mutate(Key= "eGenes (13,674)") %>% dplyr::select(Key, gencode_id) %>% distinct() %>% mutate(ensembl_gene_id=str_extract(gencode_id, regex("(ENSG[0-9]{11})")))

#5415 genes are expressed >0.1 but not spatially regulated.
expressed_non_bmap_genes <- tpm_0.1 %>% filter( !(tpm_0.1$gencode_id %in% bmap$gencode_id))  %>% mutate(Key = "non-eGenes (5,415)") %>% dplyr::select(Key, gencode_id) %>% mutate(ensembl_gene_id=str_extract(gencode_id, regex("(ENSG[0-9]{11})")))

#these now represent all genes expreesed in blood >0.1 
combined_egenes_non_egenes <- rbind (expressed_non_bmap_genes, expressed_bmap_genes)
```

#Find out biotype of each of the bmap and non bmap genes
```{r}
biotype_gencode_table <- vroom("./data/gencode.v26.annotation.gtf.gz", delim='\t', skip=5, col_names = F) %>% filter(X3=="gene" | X3=="transcript") %>% as.data.frame()

#remove the quotation marks in the dataframe:
as.data.frame(sapply(biotype_gencode_table, function(x) gsub("\"", "", x)))

#extract gencode_id and ensmbl_gene_id
biotype_gencode_table <- biotype_gencode_table %>%
  mutate(gencode_id =str_extract(X9, regex("(ENSG[0-9]{11}.[0-9])"))) %>%
  mutate(ensembl_gene_id = str_extract(X9, regex("(ENSG[0-9]{11})"))) %>% 
  mutate(biotype = str_extract(X9, regex("(gene_type.*; gene_name)"))) %>% #grep Symbol, ;Pval, and everything in between (gene name)
  mutate(biotype = gsub("gene_type \"", "", biotype)) %>% #leave only biotype name
  mutate(biotype = gsub("\"; gene_name", "",biotype)) %>% 
  dplyr::select(biotype, gencode_id, ensembl_gene_id)
```

#find biotype of non bmap genes; (all of them had biotypes) and biotypes of non bmap genes
```{r}
expressed_non_bmap_genes <- inner_join(expressed_non_bmap_genes, biotype_gencode_table, by="ensembl_gene_id") %>% distinct()

expressed_bmap_genes <- inner_join(expressed_bmap_genes, biotype_gencode_table, by="ensembl_gene_id") %>% distinct() 
```
  
#re-label biotypes to shorten list
```{r}
annotated_all_blood <- rbind(expressed_bmap_genes, expressed_non_bmap_genes) %>% dplyr::count(Key, biotype) %>% arrange(desc(n))

#find all terms with pseudogene in non-egenes
non_egenes_psudogenes <- annotated_all_blood[grep("pseudogene", annotated_all_blood$biotype),] %>% filter(Key == "non-eGenes (5,415)" ) 
#find all terms with pseudogene in egenes
egenes_psudogenes <- annotated_all_blood[grep("pseudogene", annotated_all_blood$biotype),] %>% filter(Key == "eGenes (13,674)") 

#replace different types of pseudogenes with the term "pseudogene"
annotated_all_blood <- annotated_all_blood %>%  filter( !(annotated_all_blood$biotype %in% egenes_psudogenes$biotype | annotated_all_blood$biotype %in% non_egenes_psudogenes$biotype )) %>% rbind(data.frame("biotype" = "pseudogene", "Key"= "non-eGenes (5,415)", "n"= sum(non_egenes_psudogenes$n))) %>% rbind(data.frame("biotype" = "pseudogene", "Key"= "eGenes (13,674)", "n"= sum(egenes_psudogenes$n))) 

#find all Ig and TcR genes in non-egenes
non_egenes_ig_tcr <- annotated_all_blood[grep("_gene", annotated_all_blood$biotype),] %>% filter(Key == "non-eGenes (5,415)" ) 

#find all Ig and TcR genes in egenes
egenes_ig_tcr  <- annotated_all_blood[grep("_gene", annotated_all_blood$biotype),] %>% filter(Key == "eGenes (13,674)" ) 

#replace different types of TcR genes with term "Ig variable chain and TcR genes"
annotated_all_blood <- annotated_all_blood %>%  filter( !(annotated_all_blood$biotype %in% egenes_ig_tcr$biotype | annotated_all_blood$biotype %in% non_egenes_ig_tcr$biotype )) %>% rbind(data.frame("biotype" = "Ig variable chain and TcR genes", "Key"= "non-eGenes (5,415)", "n"= sum(non_egenes_ig_tcr$n))) %>% rbind(data.frame("biotype" = "Ig variable chain and TcR genes", "Key"= "eGenes (13,674)", "n"= sum(egenes_ig_tcr$n))) 

#find all Non coding RNA genes in non-egenes 
non_egenes_ncrna <- annotated_all_blood[grep('snoRNA|ribozyme|sRNA|Mt_rRNA|Mt_tRNA|rRNA|scaRNA|scRNA|miRNA|misc_RNA|snRNA', annotated_all_blood$biotype),] %>% filter(Key == "non-eGenes (5,415)" ) 
#find all Non coding RNA genes in egenes 
egenes_ncrna  <- annotated_all_blood[grep('snoRNA|ribozyme|rRNA|scaRNA|sRNA|Mt_rRNA|Mt_tRNA|scRNA|miRNA|misc_RNA|snRNA', annotated_all_blood$biotype),] %>% filter(Key == "eGenes (13,674)" ) 

#replace different types of non coding RNA genes with term "non-coding RNA"
annotated_all_blood <- annotated_all_blood %>%  filter( !(annotated_all_blood$biotype %in% egenes_ncrna$biotype | annotated_all_blood$biotype %in% non_egenes_ncrna$biotype )) %>% rbind(data.frame("biotype" = "non coding RNA", "Key"= "non-eGenes (5,415)", "n"= sum(non_egenes_ncrna$n))) %>% rbind(data.frame("biotype" = "non coding RNA", "Key"= "eGenes (13,674)", "n"= sum(egenes_ncrna$n))) 

#other non-egenes
non_egenes_other <- annotated_all_blood[grep('non_coding|3prime_overlapping_ncRNA|sense_overlapping|bidirectional_promoter_lncRNA', annotated_all_blood$biotype),] %>% filter(Key == "non-eGenes (5,415)" ) 
#other egenes
egenes_other  <- annotated_all_blood[grep('non_coding|3prime_overlapping_ncRNA|sense_overlapping|bidirectional_promoter_lncRNA', annotated_all_blood$biotype),] %>% filter(Key == "eGenes (13,674)" ) 

#replace the remaining misc biotypes with term "other"
annotated_all_blood <- annotated_all_blood %>%  filter( !(annotated_all_blood$biotype %in% egenes_other$biotype | annotated_all_blood$biotype %in% non_egenes_other$biotype )) %>% rbind(data.frame("biotype" = "other", "Key"= "non-eGenes (5,415)", "n"= sum(non_egenes_other$n))) %>% rbind(data.frame("biotype" = "other", "Key"= "eGenes (13,674)", "n"= sum(egenes_other$n))) 
```

#Plot
```{r}
ggplot(data= annotated_all_blood, aes(x= reorder(biotype, desc(n))
, y= n, fill=
Key)) +  geom_bar(position = 'dodge', stat='identity') +
  geom_text(aes(label=n), position=position_dodge(width=0.9), vjust=-0.25, size=3) + ggtitle("genes expressed in blood >0.1 tpm") +
  theme_classic() + ylab("number of genes") + xlab("bio-type") +
  theme(plot.title = element_text(face="bold",size=9, hjust = 0.5)) + theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_fill_manual(values=c('darkcyan','gray60')) + scale_y_continuous(labels = scales::scientific) #+ scale_fill_discrete(name="Key", labels = c("eGenes (13,674)", "non-eGenes (5,415)"))
```


#plot as proportion
```{r}
#must put x and y in ggplot to prevent "xyz" requires both x and y aestitic error
FigureS1c <- ggplot(annotated_all_blood, aes(x = biotype, y = n, fill=Key)) +
  geom_bar( stat="identity", position="fill", color = "black") + theme_classic() + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + scale_x_discrete(limits=c("protein_coding","pseudogene", "antisense", "Ig variable chain and TcR genes" ,"lincRNA" , "non coding RNA" ,"processed_transcript",  "sense_intronic" ,"TEC", "other" ))  +  scale_y_continuous(labels = scales::percent) + ylab("% genes expressed in blood") + 
  geom_text(aes(label = n), position = "fill", vjust = 1.5, size=3.5) + scale_fill_manual(values=c( '#de7a52', "#f7f7f7"))
```
#Save plots 
```{r}
# pdf("./plots/FigureS1c.pdf", width =7, height = 5,  paper="USr")
# print(FigureS1c)
# dev.off()
```

#annotate expression level for each gene 
```{r}
non_egenes_ann_tpm<-merge(expressed_non_bmap_genes, tpm_0.1, by= "ensembl_gene_id") %>% filter(biotype=="protein_coding") %>%  dplyr::select(Key, `Whole Blood`) %>% mutate(Key="protein coding non-eGenes")

egenes_ann_tpm<-merge(expressed_bmap_genes, tpm_0.1, by= "ensembl_gene_id") %>% filter(biotype=="protein_coding") %>%  dplyr::select(Key, `Whole Blood`) %>% mutate(Key="protein coding eGenes")

egenes_non_egenes_ann_tpm_combined <- rbind(egenes_ann_tpm, non_egenes_ann_tpm)
```
#do stat analysis 
```{r}
egenes_non_egenes_ann_tpm_combined %>% group_by(Key) %>% dplyr::summarise(mean=mean(`Whole Blood`))
```

```{r}
egenes_non_egenes_ann_tpm_combined %>% group_by(Key) %>% dplyr::summarise(sd=sd(`Whole Blood`))
```

#Violin plot comparing protein coding egenes vs protein coding non-egenes
```{r}
FigureS1d<-ggplot(egenes_non_egenes_ann_tpm_combined, aes(x=Key, y=log10(`Whole Blood` +1), fill=Key)) + 
    geom_violin(trim=T)  + theme_minimal() + 
 stat_summary(fun.data=mean_sdl, mult=1, 
                 geom="pointrange", color="brown", shape=3, size=0.85)   + 
  labs(y = "log10(median tpm)") +
  scale_fill_brewer(palette="RdBu") + stat_summary(fun.y = "median", geom = "point", shape = 3, size = 3, color = "black") + theme(legend.position = "none") + theme(axis.text.x= element_text(size=9)) +
   stat_compare_means( method = "t.test", label = "p.signif", label.x = 1.5, label.y = 5) 

```
#Save plot
```{r}
#save as pdf
# pdf("./plots/FigureS1d.pdf", height=5, width=6, paper="USr")
# print(FigureS1d)
# dev.off()
```

#LOEUF SCORE
```{r}
constraint_table<-vroom("./data/gnomad.v2.1.1.lof_metrics.by_gene.txt", delim = "\t", , col_select = c("gene_id", "oe_lof_upper")) %>% dplyr::rename("ensembl_gene_id"=gene_id)
```

#annotate loeuf for both egenes and non-egeens
```{r}
non_egenes_ann_loeuf<-merge(expressed_non_bmap_genes, constraint_table, by= "ensembl_gene_id") %>% filter(biotype=="protein_coding") %>%  dplyr::select(Key, "oe_lof_upper") %>% mutate(Key="protein coding non-eGenes")

egenes_ann_loeuf<-merge(expressed_bmap_genes, constraint_table, by= "ensembl_gene_id") %>% filter(biotype=="protein_coding") %>%  dplyr::select(Key, "oe_lof_upper") %>% mutate(Key="protein coding eGenes")

egenes_non_egenes_ann_loeuf_combined <- rbind(non_egenes_ann_loeuf, egenes_ann_loeuf)
```
  
#stat analysis
```{r}
egenes_non_egenes_ann_loeuf_combined %>% dplyr::filter(!is.na(oe_lof_upper)) %>% group_by(Key) %>% dplyr::summarise(mean=mean(oe_lof_upper))
```

```{r}
egenes_non_egenes_ann_loeuf_combined %>% dplyr::filter(!is.na(oe_lof_upper)) %>% group_by(Key) %>% dplyr::summarise(sd=sd(oe_lof_upper))
```

#plot distribution
```{r}
FigureS1e<-ggplot(egenes_non_egenes_ann_loeuf_combined, aes(x=Key, y=oe_lof_upper, fill=Key)) + 
    geom_violin(trim=T)  + theme_minimal() + 
 stat_summary(fun.data=mean_sdl, mult=1, 
                 geom="pointrange", color="brown", shape=3, size=0.85)   + 
  labs(title="black=median, read=mean, brown line=sd", y = "LOEUF") +
  scale_fill_brewer(palette="RdBu") + stat_summary(fun.y = "median", geom = "point", shape = 3, size = 3, color = "black") + theme(legend.position = "none") + theme(axis.text.x= element_text(size=9)) +
   stat_compare_means( method = "t.test", label = "p.signif", label.x = 1.5, label.y = 2.1) 

```
#Save plot
```{r}
#save as pdf
# pdf("./plots/FigureS1e.pdf", height=5, width=6, paper="USr")
# print(FigureS1e)
# dev.off()
```


