---
title: "eQTLs_across_chr6"
author: "Roan E. Zaied"
date: "2/17/2021"
output: html_document
---

# This script was used to generate figure 3 and figure S2.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/home/rstudio/")
pdf.options(paper = "a4")
```

# load required packages
```{r}
if (!require("pacman")) install.packages("pacman")
pacman::p_load(tidyverse,ggrepel,vroom,ggpubr,dplyr,reshape2)
```

# read blood map data for chromosome 6 only
```{r}
chr6_bmap_all <- vroom("./data/blood_grn_sig.txt.gz", delim="\t") %>% dplyr::filter(snp_chr=="chr6") %>% mutate("ensembl_gene_id"=str_extract(gencode_id, regex("(ENSG[0-9]{11})"))) %>% 
  mutate(eQTL_eGene=paste0(snp, "_", ensembl_gene_id))
```

# count number of cis, trans inter and trans intra interactions per 1MB bin across chromosome 6
```{r}
interaction_type_proportion<-c()

for (i in seq(from=0, to=170000000, by= 1000000)) {

  bin_lower_bound<-i
  bin_upper_bound<-i+1000000
    #count number of snps in that bin
  snps_per_bin<- chr6_bmap_all %>% filter(chr6_bmap_all$snp_locus >= bin_lower_bound & chr6_bmap_all$snp_locus < bin_upper_bound)

  trans_intra_count<-sum(snps_per_bin$interaction_type=="Trans-intrachromosomal")
  trans_inter_count<-sum(snps_per_bin$interaction_type=="Trans-interchromosomal")
  cis_count<-sum(snps_per_bin$interaction_type=="Cis")

  #append each row to table
  interaction_type_proportion<-rbind(interaction_type_proportion, c(i, cis_count, trans_intra_count, trans_inter_count))

}

interaction_type_proportion<- interaction_type_proportion %>% as.data.frame() %>% distinct()

colnames(interaction_type_proportion) = c("bin", "Cis", "Trans_intra", "Trans_inter")
interaction_type_proportion_molten<-melt(interaction_type_proportion, id.vars="bin" )
```

# plot proportion of spatial eQTL-eGene interactions across chromosome 6 grouped by type
```{r}
Figure3a<-ggplot(interaction_type_proportion_molten) +
  geom_bar(aes(x = bin/1000000, y = value, fill=variable), stat="identity", position="fill", color = "grey70", size=0.05) + theme_minimal() + scale_x_continuous(n.breaks=28, name="Chr6 (Mb)", limits=c(0, 171)) + scale_y_continuous(name="Proportion of interactions by type") + guides(fill=guide_legend("Interaction type")) +  scale_fill_manual(values=c('white', "gray25", "cyan4"))
```

# save as pdf
```{r}
# pdf("./plots/Figure3a.pdf", height=6, width=15, paper = "USr")
# print(Figure3a)
# dev.off()
```

# functions to read chr data for specified chromosome from GTEx and tissue maps (adult brain and left ventricle)
```{r}
gtex_snps<-vroom("./data/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.lookup_table_shortened.txt.gz", delim="\t", col_select = c("rs_id_dbSNP151_GRCh38p7", "variant_pos", "chr"))

read_gtex_data<-function(chrom) {
  
gtex_chr_dt<-gtex_snps %>% filter(chr==chrom) %>% 
  mutate(group="GTEx_snps") %>% dplyr::select(-"chr", -"rs_id_dbSNP151_GRCh38p7" ) %>% 
  dplyr::rename("locus"=variant_pos, "group"=group)

return(gtex_chr_dt)
}

read_blood_data<-function(chr) {
    blood_dt<-vroom("./data/blood_grn_sig.txt.gz", delim="\t", col_select = c("snp_locus", "snp_chr")) %>% dplyr::filter(snp_chr==chr) %>% dplyr::rename("locus"=snp_locus) %>% dplyr::select(-snp_chr)

  return(blood_dt)
}

read_adult_brain_data<-function(chr) {
  brain_dt<-vroom(paste0("./data/adult_brain_cortex/", chr, "/significant_eqtls.txt"), delim="\t", col_select = c("snp_locus")) %>% 
      dplyr::rename("locus"=snp_locus)

  return(brain_dt)
}

read_left_ventricle_data<-function(chr) {
  ventricle_dt<-vroom(paste0("./data/left_ventricle/",chr,"/significant_eqtls.txt"), delim="\t", col_select = c("snp_locus")) %>% 
      dplyr::rename("locus"=snp_locus)

  return(ventricle_dt)
}
```

# read gene ref table from gtex specifying position of genes
```{r}
# remove the Y chr from table
gene_ref<-read.table("./data/gencode.v26.GRCh38.genes.gtf.gz", sep="\t") %>% 
   mutate(gene_id=str_extract(V9, regex("(ENSG[0-9]{11})"))) %>% #only use gencode number to increase match number between ref and GTEx data
  filter(V3=="gene") %>% 
  filter(!(V1=="chrY")) %>% 
  dplyr::select(V4, V5, V1) %>% 
  dplyr::rename("start"=V4, "end"=V5, "chr"=V1) %>%
  mutate("locus"=(end+start)/2) %>%  
  dplyr::select(chr, locus) %>% 
  mutate(group="density") %>% 
  distinct() #so that each gene is only counted once
```

# only select data points from chromosome 6
```{r}
gene_density_6<-gene_ref %>% 
  filter(chr=="chr6") %>% 
  dplyr::select(-chr)
```

# plot gene density across chr 6 (Fig. 3b)
```{r}
Figure3b<-ggplot(gene_density_6, aes(locus/1000000, colour = group)) +
  geom_freqpoly(binwidth = 1) + theme_minimal() + labs(x="Chr6 (Mb)", y="number of genes") +  theme(axis.text.y = element_text(size=12)) + guides(color=guide_legend("Key")) +scale_y_continuous(labels = scales::scientific) +  scale_x_continuous(n.breaks=28, limits=c(0,171)) + theme(legend.position = "none")
```

```{r}
#save as pdf
# pdf("./plots/Figure3b.pdf", height=2.2, width=7.5, paper = "USr")
# print(Figure3b)
# dev.off()
```

# read data across all tissues for chromosome 6
```{r}
gtex_chr6_snps<- read_gtex_data("chr6") 
bmap_chr6<-read_blood_data("chr6")
lventricle_chr6<-read_left_ventricle_data("chr6")
brain_chr6<-read_adult_brain_data("chr6")
gene_density_6<-gene_ref %>% 
  filter(chr=="chr6") %>% 
  dplyr::select(-chr)
```

# read data across all 3 tissues for chr19
```{r}
gtex_chr19_snps<- read_gtex_data("chr19") 
bmap_chr19<-read_blood_data("chr19")
lventricle_chr19<-read_left_ventricle_data("chr19")
brain_chr19<-read_adult_brain_data("chr19")
gene_density_19<-gene_ref %>% 
  filter(chr=="chr19") %>% 
  dplyr::select(-chr)

```

# read data across all 3 tissues for chr7
```{r}
gtex_chr7_snps<- read_gtex_data("chr7")
bmap_chr7<-read_blood_data("chr7")
lventricle_chr7<-read_left_ventricle_data("chr7")
brain_chr7<-read_adult_brain_data("chr7")
gene_density_7<-gene_ref %>% 
  filter(chr=="chr7") %>% 
  dplyr::select(-chr)

```

# read data across all 3 tissues for chr18
```{r}
gtex_chr18_snps<- read_gtex_data("chr18")
bmap_chr18<-read_blood_data("chr18")
lventricle_chr18<-read_left_ventricle_data("chr18")
brain_chr18<-read_adult_brain_data("chr18")
gene_density_18<-gene_ref %>% 
  filter(chr=="chr18") %>% 
  dplyr::select(-chr)
```

# function to bin chromosome into 1Mb bins and compute ratio of blood map eQTL-eGene interaction to SNPs 
```{r}
compute_count_and_ratio<-function(chr_tissue_map, gtex_chr_dt) {
tissue_map_count_combined<-c()

#for each bin
for (i in seq(from=0, to=max(gtex_chr_dt$locus), by= 1000000)) {
  
  bin_lower_bound<-i
  bin_upper_bound<-i+1000000
    #count number of snps in that bin. < and not <= since <= is counted in next bin 
  gtex_count_per_bin<-sum(gtex_chr_dt$locus >= bin_lower_bound & gtex_chr_dt$locus < bin_upper_bound)
  tissue_map_count_per_bin<-sum(chr_tissue_map$locus >= bin_lower_bound & chr_tissue_map$locus < bin_upper_bound)
  
 ratio<-tissue_map_count_per_bin/gtex_count_per_bin
  
  #append each row to table
  tissue_map_count_combined<-rbind(tissue_map_count_combined, c(i, tissue_map_count_per_bin, gtex_count_per_bin, ratio))
  
}

tissue_map_count_combined<- tissue_map_count_combined %>% as.data.frame() %>%
  dplyr::rename("bin"=V1, "ma2p_eqtl_count"=V2, "GTEx_snps"=V3, "eQTLs_eGene/snps_ratio"=V4) %>% 
  distinct() #because rmd is v. annoying and repeats scripts twice 

return(tissue_map_count_combined)
}
```

# find ratio of blood GRN eQTL-eGene interaction to SNPs for chromosome 6 in all 3 tissue maps
```{r}
bmap_count_combined_chr6<-compute_count_and_ratio(bmap_chr6, gtex_chr6_snps) %>% 
  mutate(group="whole blood")

adult_brain_count_combined_chr6<-compute_count_and_ratio(brain_chr6, gtex_chr6_snps) %>%
  mutate(group="adult brain cortex")

left_ventricle_count_combined_chr6<-compute_count_and_ratio(lventricle_chr6, gtex_chr6_snps) %>% 
  mutate(group="left ventricle") 

combined_ratio_chr6<-rbind(left_ventricle_count_combined_chr6, bmap_count_combined_chr6) %>% 
  rbind(adult_brain_count_combined_chr6) %>% replace(is.na(.), 0) 

```

# plot Ratio of number sig. eQTLs-egene pairs and SNPs (whole blood) across different tissues in chr6
```{r}
Figure3c<-ggplot(combined_ratio_chr6, aes(x=bin/1000000, y=`eQTLs_eGene/snps_ratio`, color=group)) +
  geom_line(binwidth = 1) + theme_minimal() + labs(x="Chr6 (Mb)", y="ratio (eQTL-eGene interactions/SNPs)") + guides(color=guide_legend("Tissue")) + theme(axis.text.y = element_text(size=10), legend.text=element_text(size=10)) + scale_color_manual(values=c("#7CAE00", "#00BFC4", "#f8766D")) + scale_x_continuous(n.breaks=28, limits=c(0,171)) 
```

```{r}
#save as pdf
# pdf("./plots/Figure3c.pdf", height=2.5, width=7.5, paper = "USr")
# print(Figure3c)
# dev.off()
```

# plot Ratio of number sig. eQTLs-egene pairs and SNPs (whole blood) across different tissues in chr6 (HLA region only)
```{r}
Figure3d<-ggplot(combined_ratio_chr6, aes(x=bin/1000000, y=`eQTLs_eGene/snps_ratio`, color=group)) +
  geom_line(binwidth = 1) + theme_classic() + labs(x="Chr6 (Mb)", y="ratio (eQTL-eGene interactions/SNPs)") + guides(color=guide_legend("Tissue")) +  scale_x_continuous(limits=c(25,34), n.breaks=19) +scale_color_manual(values=c("#7CAE00", "#00BFC4", "#f8766D")) + theme(axis.text.x= element_text(size=10))
```

```{r}
#save as pdf
# pdf("./plots/Figure3d.pdf", height=2.5, width=7.5, paper = "USr")
# print(Figure3d)
# dev.off()
```


# find ratio of blood GRN eQTL-eGene interaction to SNPs for chromosome 7 in all 3 tissue maps and plot (Fig. S2a)
```{r}
bmap_count_combined_chr7 <-compute_count_and_ratio(bmap_chr7, gtex_chr7_snps) %>% 
  mutate(group="whole blood")

adult_brain_count_combined_chr7<-compute_count_and_ratio(brain_chr7, gtex_chr7_snps) %>% 
  mutate(group="adult brain cortex")

left_ventricle_count_combined_chr7<-compute_count_and_ratio(lventricle_chr7, gtex_chr7_snps) %>% 
  mutate(group="left ventricle")

combined_ratio_chr7<-rbind(left_ventricle_count_combined_chr7) %>% 
  rbind(adult_brain_count_combined_chr7) %>% 
  rbind(bmap_count_combined_chr7)%>% replace(is.na(.), 0) #bins having 0/0 will produce NAs
 
FigureS2a<-ggplot(combined_ratio_chr7, aes(x=bin/1000000,y=`eQTLs_eGene/snps_ratio`, color=group)) +
geom_line() +
  theme_minimal() + labs(x="Chr7 (Mb)", y="ratio (eQTL-eGene interactions/SNPs)") +scale_color_manual(values=c("#7CAE00", "#00BFC4", "#f8766D")) + scale_color_discrete(name= "Tissue", labels = c("whole blood", "adult brain cortex", "left ventricle"))
```

```{r}
#save as pdf
# pdf("./plots/FigureS2a.pdf", height=2.5, width=7.5, paper = "USr")
# print(FigureS2a)
# dev.off()

```


# find ratio of blood GRN eQTL-eGene interaction to SNPs for chromosome 18 in all 3 tissue maps and plot (Fig. S2b)
```{r}
bmap_count_combined_chr18<-compute_count_and_ratio(bmap_chr18, gtex_chr18_snps) %>% 
  mutate(group="whole blood")

adult_brain_count_combined_chr18<-compute_count_and_ratio(brain_chr18, gtex_chr18_snps)%>% 
  mutate(group="adult brain cortex")

left_ventricle_count_combined_chr18<-compute_count_and_ratio(lventricle_chr18, gtex_chr18_snps) %>% 
  mutate(group="left ventricle")

combined_ratio_chr18<-rbind(left_ventricle_count_combined_chr18) %>% 
  rbind(adult_brain_count_combined_chr18) %>% 
  rbind(bmap_count_combined_chr18) %>% replace(is.na(.), 0) 

FigureS2b<-ggplot(combined_ratio_chr18, aes(x=bin/1000000,y=`eQTLs_eGene/snps_ratio`, color=group)) +
geom_line() +
  theme_minimal() + labs(x="Chr18 (Mb)", y="ratio (eQTL-eGene interactions/SNPs)") + scale_color_manual(values=c("#7CAE00", "#00BFC4", "#f8766D")) + scale_color_discrete(name= "Tissue", labels = c("whole blood", "adult brain cortex", "left ventricle"))
```

```{r}
#save as pdf
# pdf("./plots/FigureS2b.pdf", height=2.5, width=7.5, paper = "USr")
# print(FigureS2b)
# dev.off()
```

#find ratio of blood GRN eQTL-eGene interaction to SNPs for chromosome 19 in all 3 tissue maps and plot (Fig. S2c)
```{r}
bmap_count_combined_chr19 <-compute_count_and_ratio(bmap_chr19, gtex_chr19_snps) %>% 
  mutate(group="whole blood")

adult_brain_count_combined_chr19<-compute_count_and_ratio(brain_chr19, gtex_chr19_snps) %>%
  mutate(group="adult brain cortex")

left_ventricle_count_combined_chr19<-compute_count_and_ratio(lventricle_chr19, gtex_chr19_snps) %>% 
  mutate(group="left ventricle") 

combined_ratio_chr19<-rbind(left_ventricle_count_combined_chr19, bmap_count_combined_chr19) %>% 
  rbind(adult_brain_count_combined_chr19) %>% replace(is.na(.), 0) 
   
FigureS2c<-ggplot(combined_ratio_chr19, aes(x=bin/1000000,y=`eQTLs_eGene/snps_ratio`, color=group)) +
geom_line() +
  theme_minimal() + labs( x="Chr19 (Mb)", y="ratio (eQTL-eGene interactions/SNPs)") + scale_color_manual(values=c("#7CAE00", "#00BFC4", "#f8766D")) + scale_color_discrete(name= "Tissue", labels = c("whole blood", "adult brain cortex", "left ventricle")) 
```

```{r}
#save as pdf
# pdf("./plots/FigureS2c.pdf", height=2.5, width=7.5, paper = "USr")
# print(FigureS2c)
# dev.off()
```
